crates/ghost_actor/src/ghost_system.rs

```diff
@@ -0,0 +1,284 @@
+use crate::*;
+use std::sync::Arc;
+
+/// struct used for hinting on whether / when to next run this process fn
+pub struct GhostProcessInstructions {
+    should_continue: bool,
+    next_run_delay_ms: u64,
+}
+
+impl Default for GhostProcessInstructions {
+    fn default() -> Self {
+        Self {
+            should_continue: false,
+            next_run_delay_ms: 0,
+        }
+    }
+}
+
+impl GhostProcessInstructions {
+    pub fn should_continue(&self) -> bool {
+        self.should_continue
+    }
+
+    pub fn set_should_continue(mut self, should_continue: bool) -> Self {
+        self.should_continue = should_continue;
+        self
+    }
+
+    pub fn next_run_delay_ms(&self) -> u64 {
+        self.next_run_delay_ms
+    }
+
+    pub fn set_next_run_delay_ms(mut self, next_run_delay_ms: u64) -> Self {
+        self.next_run_delay_ms = next_run_delay_ms;
+        self
+    }
+}
+
+/// typedef for a periodic process callback
+pub type GhostProcessCb<'lt> =
+    Box<dyn FnMut() -> GhostResult<GhostProcessInstructions> + 'lt + Send + Sync>;
+
+/// internal struct for tracking processor fns
+struct GhostProcessorData<'lt> {
+    pub delay_until: Option<std::time::Instant>,
+    pub cb: GhostProcessCb<'lt>,
+}
+
+/// inner struct trackes queued process callbacks
+struct GhostSystemInner<'lt> {
+    process_recv: crossbeam_channel::Receiver<GhostProcessorData<'lt>>,
+    process_queue: Vec<GhostProcessorData<'lt>>,
+}
+
+impl<'lt> GhostSystemInner<'lt> {
+    /// new inner system
+    fn new(process_recv: crossbeam_channel::Receiver<GhostProcessorData<'lt>>) -> Self {
+        Self {
+            process_recv,
+            process_queue: Vec::new(),
+        }
+    }
+
+    /// first, check for new process functions,
+    /// then, actually loop through the queued process functions
+    /// keep them for next time if they return true
+    fn process(&mut self) -> GhostResult<()> {
+        while let Ok(item) = self.process_recv.try_recv() {
```

Comment:
Is there a chance this won't terminate? If so we should cut this off at some finite limit.

Reply:
It also suffers from making the queue arbitrarily large.

Reply:
mmm... As far as the system is concerned, any process callback waiting in this queue *must* be processed, or the system won't function properly.

As far as code errors, like infinite loops that create ghost actors, etc... I was thinking it might be interesting at some point to have some configurable "warning" level metrics, like...

- if there are > 1,000,000 actors in the system, or
- the last process() was called > 100ms ago, etc